crates/hir-ty/src/traits.rs

```diff
@@ -90,6 +94,37 @@ pub(crate) fn normalize_projection_query(
     projection: ProjectionTy,
     env: Arc<TraitEnvironment>,
 ) -> Ty {
+    struct InferenceVarVisitor(Option<()>);
+    impl TypeVisitor<Interner> for InferenceVarVisitor {
+        type BreakTy = ();
+
+        fn as_dyn(&mut self) -> &mut dyn TypeVisitor<Interner, BreakTy = Self::BreakTy> {
+            self
+        }
+
+        fn interner(&self) -> Interner {
+            Interner
+        }
+
+        fn visit_inference_var(
+            &mut self,
+            _var: chalk_ir::InferenceVar,
+            _outer_binder: DebruijnIndex,
+        ) -> std::ops::ControlFlow<Self::BreakTy> {
+            self.0 = Some(());
+            std::ops::ControlFlow::Break(())
+        }
+    }
+
+    let mut visitor = InferenceVarVisitor(None);
+    projection.visit_with(visitor.as_dyn(), DebruijnIndex::INNERMOST);
+    if visitor.0.is_some() {
```

Comment:
this could be a function `has_inference_variables`. Also, the `flags` in a `Ty` already store whether it contains any inference variable, so it could be easier to use that :)